pkg/k8s: fix all structural issues with CNP validation

aanm opened the pull request:
CNP schema validation was incorrectly formatted for some fields which
could cause badly formatted yaml files to be accepted by kube-apiserver
bypassing the schema validation. This would later cause Cilium to print
errors and potentially avoid it from starting, as the invalid CNPs would
prevent Cilium from fully synchronize with kube-apiserver an operation
that is essential when starting Cilium.

This commit fixes all violations presented by kubernetes for the CCNP
and CNP validation which would then deny bad CNPs and / or CCNPs from
being accepted by kube-apiserver.

Signed-off-by: André Martins <andre@cilium.io>

Fixes: #10643

```release-note
Tight CNP and CCNP schema validation for badly formatted policies (yaml or json) 
```

aanm committed:
pkg/k8s: fix all structural issues with CNP validation

CNP schema validation was incorrectly formatted for some fields which
could cause badly formatted yaml files to be accepted by kube-apiserver
bypassing the schema validation. This would later cause Cilium to print
errors and potentially avoid it from starting, as the invalid CNPs would
prevent Cilium from fully synchronize with kube-apiserver an operation
that is essential when starting Cilium.

This commit fixes all violations presented by kubernetes for the CCNP
and CNP validation which would then deny bad CNPs and / or CCNPs from
being accepted by kube-apiserver.

Signed-off-by: André Martins <andre@cilium.io>

joestringer commented on pkg/k8s/apis/cilium.io/v2/register.go:
```diff
@@ -854,19 +857,29 @@ var (
+		Type: "object",
 		Properties: map[string]apiextensionsv1beta1.JSONSchemaProps{
-			"SecurityGroupsIds": {
+			"securityGroupsIds": {
```
Is there any meaningful potential impact with the case change here or are they seen as the same thing when upgrading schemas?

aanm committed:
add validation checker for installed CNPs in preflight

As Cilium might require to update its CRD validation schema it is
important for the users to make sure all policies installed in their
cluster are valid in the point of view the new CRD validation schema
before performing an upgrade.

Avoiding doing this validation might cause Cilium from updating its
NodeStatus in those invalid Network Policies as well as in the worst
case scenario it might give a false sense of security to the user if
a policy is badly formatted and Cilium is not inforcing that policy
due a bad validation.

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
examples/crds: update CRD with the latest schema

Signed-off-by: André Martins <andre@cilium.io>

aanm committed:
test: fix badly formatted CNP

Signed-off-by: André Martins <andre@cilium.io>

tgraf commented on Documentation/install/upgrade.rst:
```diff
@@ -1145,3 +1161,109 @@ If a migration has gone wrong, it possible to start with a clean slate. Ensure t
+
+Running the CNP Validator will make sure the policies deployed in the cluster
+are valid. It is important to run this validation before an upgrade so it will
+make sure Cilium has a correct behavior after upgrade. This CNP Validator is
```
Add a sentence why this is required

tgraf commented on cilium/cmd/preflight_k8s_valid_cnp.go:
```diff
@@ -0,0 +1,140 @@
+	Short: "Validate Cilium Network Policies deployed in the cluster",
+	Long: `Before upgrading Cilium it is recommended to run this validation checker
+to make sure the policies deployed are valid. The validator will verify if all policies
+deployed in the cluster all valid, in case they are not an error is printed and the
```
s/all/are/
... in case they are not, an error

tgraf commented on cilium/cmd/preflight_k8s_valid_cnp.go:
```diff
@@ -0,0 +1,140 @@
+	Long: `Before upgrading Cilium it is recommended to run this validation checker
+to make sure the policies deployed are valid. The validator will verify if all policies
+deployed in the cluster all valid, in case they are not an error is printed and the
+has an exit code -1.`,
```
exit code -1 is returned.

tgraf commented on cilium/cmd/preflight_k8s_valid_cnp.go:
```diff
@@ -0,0 +1,140 @@
+	// the CLI tool.
+	logging.DefaultLogger.SetFormatter(log.Formatter)
+
+	log.Info("Setting up kubernetes client")
```
_K_ubernetes

joestringer commented on cilium/cmd/preflight_k8s_valid_cnp.go:
```diff
@@ -0,0 +1,140 @@
+		log.WithError(err).Fatal("Unable to connect to Kubernetes apiserver")
+	}
+
+	ctx, initCancel := context.WithTimeout(context.Background(), validateK8sPoliciesTimeout)
```
Should we have a real timeout here, and for the other contexts around this area?

aanm replied:
I don't think so because this is a user request and we don't know the number of CNPs installed to have a timeout here. Also, this will be triggered by the user so the user can cancel the request with CTRL+C

joestringer replied:
I'm thinking from the perspective of this being run automatically as part of the preflight. What if some issue causes APIserver connectivity breakage then, will the pod just hang in not-ready state? How will users know?

joestringer reviewed (approved):
If a user specifies the APIserver in Helm (because of kube-proxy-free), will this tool automatically pick that up and use it?

I looked through, nothing looked overly out-of-place but I'm not the best reviewer for the k8s side.

Upgrade-wise I guess we're basically saying that the best time to validate your policy is before you move to the new version with the stricter validation?

One more question, is there an easy way for me to just run this on my cluster? Just exec into the cilium pod and run it there? (Is that advisable?) Let's say I wanted to validate my policies but without the rest of the upgrade process.

The new tool should be added to k8s code owners.

aanm reviewed (commented):
There was a typo in the docs:
```patch
diff --git b/Documentation/install/upgrade.rst a/Documentation/install/upgrade.rst
index fb35dabae..2d4d02bac 100644
--- b/Documentation/install/upgrade.rst
+++ a/Documentation/install/upgrade.rst
@@ -66,7 +66,7 @@ file.
         --set config.enabled=false \\
         --set operator.enabled=false \\
         --set global.k8sServiceHost=API_SERVER_IP \\
-        --set global.k8sServicePort=API_SERVER_PORT
+        --set global.k8sServicePort=API_SERVER_PORT \\
         > cilium-preflight.yaml
       kubectl create cilium-preflight.yaml
```

I've fix it. Will merge the PR once the GH check for docs have passed. All other tests have passed before this typo was fixed.
